spicy/toolchain/src/compiler/codegen/parsers/literals.cc

```diff
@@ -56,42 +56,61 @@ struct Visitor : public visitor::PreOrder {
     }
     auto popBuilder() { return pb()->popBuilder(); }
 
+    auto needToCheckForLookAhead(const Meta& meta) {
+        bool needs = false;
+
+        if ( auto field = lp->production->meta().field(); field && field->attributes()->find("&synchronize") )
+            needs = true;
+        else {
+            auto tokens = pb()->cg()->astInfo().look_aheads_in_use;
+            needs = tokens.find(lp->production->tokenID()) != tokens.end();
+        }
+
+        if ( pb()->options().debug && ! needs )
+            builder()->addAssert(builder()->not_(state().lahead), "unexpected look-ahead token pending", meta);
+
+        return needs;
+    }
+
     void operator()(hilti::ctor::Bytes* n) final {
-        auto error_msg = fmt("expecting '%s'", n->value());
         auto len = builder()->integer(static_cast<uint64_t>(n->value().size()));
-        auto cond = builder()->memberCall(state().cur, "starts_with", {builder()->expression(n)});
 
         switch ( state().literal_mode ) {
             case LiteralMode::Default:
             case LiteralMode::Skip: {
-                auto [have_lah, no_lah] = builder()->addIfElse(state().lahead);
+                bool check_for_look_ahead = needToCheckForLookAhead(n->meta());
+                if ( check_for_look_ahead ) {
+                    auto [have_lah, no_lah] = builder()->addIfElse(state().lahead);
 
-                pushBuilder(have_lah);
+                    pushBuilder(have_lah);
 
-                pushBuilder(builder()->addIf(
-                    builder()->unequal(state().lahead, builder()->integer(lp->production->tokenID()))));
-                pb()->parseError("unexpected token to consume", n->meta());
-                popBuilder();
+                    pushBuilder(builder()->addIf(
+                        builder()->unequal(state().lahead, builder()->integer(lp->production->tokenID()))));
+                    pb()->parseError("unexpected token to consume", n->meta());
+                    popBuilder();
 
-                pushBuilder(builder()->addIf(
-                    builder()->unequal(builder()->expression(n),
-                                       builder()->memberCall(state().cur, "sub",
-                                                             {builder()->begin(state().cur), state().lahead_end}))));
-                pb()->parseError("unexpected data when consuming token", n->meta());
-                popBuilder();
+                    pushBuilder(
+                        builder()->addIf(builder()->unequal(builder()->expression(n),
```

Comment:
Less readable, and not convinced performance is better, depends on what the C++ compiler does with this code.

Reply:
My suspicion was that it would for sure need to allocate an additional temporary for the function call. For a small sample this doesn't seem to be the case ([assembler output from clang-19.0.1](https://godbolt.org/z/KaxEYcGMq)) and instead it automatically reuses the temporary; it still does reset its value after the call though.